assembly/memory.ts

```diff
@@ -269,45 +303,113 @@ export class Memory {
     return res.fault;
   }
 
-  private getChunks(access: Access, address: u32, bytes: u8): Chunks {
+  bytesRead(address: u32, destination: Uint8Array): MaybePageFault {
+    let nextAddress = address;
+    let destinationIndex = 0;
+
+    while (destinationIndex < destination.length) {
+      const bytesLeft = destination.length - destinationIndex;
+      const pageData = this.getPage(Access.Read, nextAddress);
+      if (pageData.fault.isFault) {
+        return pageData.fault;
+      }
+      const relAddress = pageData.relativeAddress;
+      const bytesToRead = relAddress + bytesLeft < PAGE_SIZE ? bytesLeft : PAGE_SIZE - pageData.relativeAddress;
+      // actually copy the bytes
+      const source = pageData.page.raw.data.subarray(relAddress, relAddress + bytesToRead);
+      destination.set(source, destinationIndex);
+      // move the pointers
+      destinationIndex += bytesToRead;
+      nextAddress += bytesToRead;
+    }
+
+    return NO_PAGE_FAULT;
+  }
+
+  bytesWrite(address: u32, source: Uint8Array): MaybePageFault {
+    let nextAddress = address;
+    let sourceIndex = 0;
+
+    while (sourceIndex < source.length) {
+      const bytesLeft = source.length - sourceIndex;
+      const pageData = this.getPage(Access.Write, nextAddress);
```

Comment:
@coderabbitai Create an issue to handle a case where we write data on multiple pages and page fault at the end. Such behaviour should not alter the memory at all, while currently it will successfully write to all pages prior to an error.